apps/rush-lib/src/logic/buildCache/AzureStorageBuildCacheProvider.ts

```diff
@@ -0,0 +1,378 @@
+      },
+      { authorityHost: authorityHost }
+    );
+    const blobServiceClient: BlobServiceClient = new BlobServiceClient(
+      this._storageAccountUrl,
+      deviceCodeCredential
+    );
+
+    const startsOn: Date = new Date();
+    const expires: Date = new Date(Date.now() + SAS_TTL);
+    const key: ServiceGetUserDelegationKeyResponse = await blobServiceClient.getUserDelegationKey(
+      startsOn,
+      expires
+    );
+
+    const blobSasPermissions: BlobSASPermissions = new BlobSASPermissions();
+    blobSasPermissions.read = true;
+    blobSasPermissions.create = this._isCacheWriteAllowed;
+
+    const userDelegationKey: UserDelegationKey = key;
+    const queryParameters: SASQueryParameters = generateBlobSASQueryParameters(
+      {
+        startsOn: startsOn,
+        expiresOn: expires,
+        permissions: blobSasPermissions,
+        containerName: this._storageContainerName,
+        blobName: 'dummy-blob-name'
+      },
+      userDelegationKey,
+      this._storageAccountName
+    );
+
+    return queryParameters;
+  }
+
+  private _printMessageInBox(message: string, terminal: Terminal): void {
+    const boxWidth: number = Math.floor(Utilities.getConsoleWidth() / 2);
+    const maxLineLength: number = boxWidth - 10;
+
+    const wrappedMessage: string = Utilities.wrapWords(message, maxLineLength);
+    const wrappedMessageLines: string[] = wrappedMessage.split('\n');
+
+    // ╔═══════════╗
+    // ║  Message  ║
+    // ╚═══════════╝
+    terminal.writeLine(' ╔' + new Array(boxWidth - 3).join('═') + '╗ ');
+    for (const line of wrappedMessageLines) {
+      const trimmedLine: string = line.trim();
+      const padding: number = boxWidth - trimmedLine.length - 4;
+      const leftPadding: number = Math.floor(padding / 2);
+      const rightPadding: number = padding - leftPadding;
+      terminal.writeLine(
+        ' ║' +
+          new Array(leftPadding + 1).join(' ') +
+          trimmedLine +
+          (new Array(rightPadding + 1).join(' ') + '║ ')
+      );
+    }
+    terminal.writeLine(' ╚' + new Array(boxWidth - 3).join('═') + '╝ ');
+  }
+
+  private _getSasStringFromQueryParameters(sasQueryParameters: SASQueryParameters): string {
+    const sasQuerySearchParameters: URLSearchParams = new URLSearchParams();
+    for (const [parameterName, parameterValue] of Object.entries(sasQueryParameters)) {
+      if (parameterValue) {
+        let serializedParameterValue: string;
+        if (parameterValue instanceof Date) {
+          serializedParameterValue = parameterValue.toISOString();
+        } else {
+          serializedParameterValue = parameterValue;
+        }
+        sasQuerySearchParameters.append(parameterName, serializedParameterValue);
+      }
+    }
+
+    return sasQuerySearchParameters.toString();
+  }
+
+  private _getConnectionString(sasString: string | undefined): string {
+    const blobEndpoint: string = `BlobEndpoint=${this._storageAccountUrl}`;
```

Comment:
Can the URL contain  a semicolon?  If so, does it need to be escaped somehow?

(Ideally the Azure SDK would already provide an API for correctly encoding connection strings.)

Reply:
No. We generate that URL from the storage account name.